### PR TITLE
Added abstract base model for collection stats for time based averages

### PR DESCRIPTION
**Pull request by jpaidoussi**

### CODE REVIEW

**Comment by jpaidoussi** [comment]
@rudigiesler Ready for review

File: casepro/statistics/models.py [hunk 1]
```diff
@@ -122,6 +124,30 @@ class Meta:
         abstract = True
 
 
+class BaseAverage(BaseCount):
+    """
+    Tracks total minute counts of different items (e.g. time since assigned ) in different scopes (e.g. org, user)
```
Comment (rudigiesler) [inline casepro/statistics/models.py]: Shouldn't this be `average`, since the goal of this class is to track averages?

Comment (jpaidoussi) [inline casepro/statistics/models.py]: The average isn't stored directly but computer from total / count. The wording could be better though.

File: casepro/statistics/models.py [hunk 2]
```diff
@@ -122,6 +124,30 @@ class Meta:
...
+
+    squash_sql = """
+        WITH removed as (
+            DELETE FROM %(table_name)s WHERE %(delete_cond)s RETURNING "count", "total"
+        )
+        INSERT INTO %(table_name)s(%(insert_cols)s, "count", "total")
+        VALUES (
+            %(insert_vals)s,
+            GREATEST(0, (SELECT SUM("count") FROM removed)),
+            GREATEST(0, (SELECT SUM("total") FROM removed))
```
Comment (rudigiesler) [inline casepro/statistics/models.py]: It's possible that we might be wanting to track negative averages. Perhaps we should use `COALESCE` here instead.

**Comment by rudigiesler** [comment]
Just those two comments from me, looking good so far. We probably want to add some helper functions for adding values and getting averages, but those are probably better suited for the child classes like `DailyAverage`.

**Comment by jpaidoussi** [comment]
@rudigiesler Update and ready for review again.

**Comment by jpaidoussi** [comment]
@rowanseymour This is the PR for the work mentioned the dev group for adding a new stat base class for tracking time totals that we will use to calculate averages.

**Comment by rudigiesler** [comment]
I realised that this sql code is actually something that we can do in Django:
```
with transaction.atomic():
    queryset = cls.objects.filter(**{f: unsquashed[f] for f in cls.squash_over})
    unsquashed['count'] = queryset.aggregate(Sum('count'))['count__sum']
    queryset.delete()
    cls.objects.create(**unsquashed)
```
But rather than change things (we would have to check the performance of a transaction), I think it's better for now to just carry on with how we are currently doing it, I just wanted to list it here for the future.

This PR looks good to me, :+1: and 🍰 for the good custom SQL. I'm assuming that tests for these things will come when we implement the specific child classes (same as the existing `Count` things)

**Comment by rowanseymour** [comment]
@rudigiesler I'm not sure that works. The rows aggregated over won't necessarily be the same rows that are deleted. Maybe if you were able to set the transaction isolation level to SERIALIZABLE but that comes with performance implications.

**Comment by rudigiesler** [comment]
@rowanseymour Yes you're correct. The SQL probably gives us enough of a performance boost over a SERIALIZABLE transaction that it's worth it.

**Pull request merged**